kube-dns panic in treecache

(blocking #27819)
This is in a modified local copy, but I don't think I've introduced this.

```
I0623 01:36:31.861596       1 dns.go:422] Received DNS Request:federated-service.e2e-tests-service-p27lt.federation.default.svc.cluster.local., exact:true                                    
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                       
[signal 0xb code=0x1 addr=0x8 pc=0x64208c]                                                                                                                                                    

goroutine 114 [running]:
panic(0x1756500, 0xc820014060)
/usr/local/go/src/runtime/panic.go:481 +0x3e6                                                                                                                                                 
k8s.io/kubernetes/pkg/dns.(*TreeCache).getEntry(0xc820014ec0, 0xc82022e5f0, 0x11, 0xc820254150, 0x6, 0x7, 0x0, 0x0, 0x6400)                                                                   
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/dns/treecache.go:101 +0x6c                                                                                          
k8s.io/kubernetes/pkg/dns.(*KubeDNS).Records(0xc820116000, 0xc82022e5f0, 0x4f, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/dns/dns.go:442 +0x60b
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server.(*server).CNAMERecords(0xc82000a600, 0xc82022e5f0, 0x4f, 0xe00010005, 0xc82022e5f0, 0x4f, 0x0, 0x0, 0x0, 0x0, ...)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server/server.go:658 +0xa4                                                      
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server.(*server).ServeDNS(0xc82000a600, 0x7f74497fee50, 0xc820261d80, 0xc8202a2360)                                                 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server/server.go:375 +0x329a                                                    
k8s.io/kubernetes/vendor/github.com/miekg/dns.(*ServeMux).ServeDNS(0xc820015430, 0x7f74497fee50, 0xc820261d80, 0xc8202a2360)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:209 +0xbb
k8s.io/kubernetes/vendor/github.com/miekg/dns.(*Server).serve(0xc8200524e0, 0x7f74498500e8, 0xc8202a10b0, 0x7f74498500c0, 0xc820015430, 0xc82023fc00, 0x60, 0x200, 0xc82014a028, 0xc820299b60, ...)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:576 +0x7ef                                                                        
created by k8s.io/kubernetes/vendor/github.com/miekg/dns.(*Server).serveUDP                                                                                                                   
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:530 +0x3e1                                                                        
```

Yeah, this is also present in 1.3.  Here's a stack trace from that.

```
I0623 02:17:49.283528       1 dns.go:421] Received DNS Request:federated-service.e2e-tests-service-p27lt.federation.default.svc.cluster.local., exact:true
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0x63fa0c]                                                                                                                                                    
 
goroutine 94 [running]:                                                                                                                                                                       
panic(0x16ddd60, 0xc820014060)
/usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/dns.(*TreeCache).getEntry(0xc8204c0a00, 0xc8204aff40, 0x11, 0xc82035dc70, 0x6, 0x7, 0x0, 0x0, 0x6400)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/dns/treecache.go:101 +0x6c                                                                                          
k8s.io/kubernetes/pkg/dns.(*KubeDNS).Records(0xc8201b5080, 0xc8204aff40, 0x4f, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)                                                                                  
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/dns/dns.go:441 +0x60b                                                                                               
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server.(*server).CNAMERecords(0xc8204d0280, 0xc8204aff40, 0x4f, 0xe00010005, 0xc8204aff40, 0x4f, 0x0, 0x0, 0x0, 0x0, ...)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server/server.go:656 +0xa4
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server.(*server).ServeDNS(0xc8204d0280, 0x7f1511831528, 0xc8204cfe80, 0xc8202693b0)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server/server.go:375 +0x329a                                                    
k8s.io/kubernetes/vendor/github.com/miekg/dns.(*ServeMux).ServeDNS(0xc820014070, 0x7f1511831528, 0xc8204cfe80, 0xc8202693b0)                                                                  
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:209 +0xbb                                                                         
k8s.io/kubernetes/vendor/github.com/miekg/dns.(*Server).serve(0xc8200520d0, 0x7f1511800618, 0xc820267a10, 0x7f15118005f0, 0xc820014070, 0xc820276000, 0x60, 0x200, 0xc8200c0000, 0xc8202745e0, ...)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:576 +0x7ef
created by k8s.io/kubernetes/vendor/github.com/miekg/dns.(*Server).serveUDP
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:530 +0x3e1                                                                        
```

--- a/pkg/dns/treecache.go
+++ b/pkg/dns/treecache.go
@@ -98,6 +98,9 @@ func (cache *TreeCache) setSubCache(key string, subCache *TreeCache, path ...str
 
 func (cache *TreeCache) getEntry(key string, path ...string) (interface{}, bool) {
 	childNode := cache.getSubCache(path...)
+	if childNode == nil {
+		return nil, false
+	}
 	val, ok := childNode.Entries[key]
 	return val, ok
 }

